### PR TITLE
Create a Mixed Reality Toolkit Core NPM package, required dependency of the Foundation

### PR DESCRIPTION
**Pull request by davidkline-ms**
This change addresses feedback from customers and partners who would like an NPM package containing the code needed to create custom systems, without the default implementations.

This new package contains the following folders, previously packaged as part of the Foundation:

- Core \
  - Attributes
  - Definitions
  - EventDatum
  - Extensions
  - Inspectors
  - Interfaces
  - Providers
  - Services
  - Utilities

The new package includes: interfaces, data types, utility functionality, base classes, the Mixed Reality Toolkit object and the MRTK Configuration window.

The Foundation package takes a dependency on the Core.

Other changes include updates to the descriptions of the packages.

### CODE REVIEW

**Comment by djee-ms** [comment]
It is certainly a step in the right direction, and much appreciated. But I feel it could go even further. A quick `dir /s` shows 1415 files in that folder. Accounting for `.meta` and other stuffs, that's maybe 500 to 700 scripts, so really a lot of things not necessarily related to each other.

The two main systems that I hear frequently mentioned for reuse are the service singleton system, to work around Unity's arbitrary script execution order, and the Unity build system to run Unity tests on CI (we've discussed that one with 
@MenelvagorMilsom), which really are unrelated to Mixed Reality, and could even be shared and reused for any Unity project beyond the boundaries of MR apps. There are probably other pieces of interest too in the same category.

Is there a plan, probably beyond this PR, to make an inventory of all the pieces of Core and further split them down into self-contained packages?

**Comment by Railboy** [comment]
This sounds great.

>and the MRTK Configuration window

Consider leaving this out. If the goal is to deliver a super-lightweight package then a popup window will rub folks the wrong way. Also most of the settings relate to rendering / art assets, which likely won't be relevant to devs taking a core dependency just for the service registry.

**Comment by keveleigh** [comment]
> to work around Unity's arbitrary script execution order

To be fair...it isn't really arbitrary. If that's all you're looking to do, you're probably better off explicitly setting it at 
image

File: Assets/MRTK/Tests/TestUtilities/packagetemplate.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "msftFeatureCategory": "Mixed Reality Toolkit",
     "version": "%version%",
     "unity": "2019.4",
-    "description": "Test utilities for use with the Microsoft Mixed Reality Toolkit",
+    "description": "TUutilities for use are a set of scripts that accelerate writing tests for features built using the Microsoft Mixed Reality Toolkit",
```
Comment (keveleigh) [inline Assets/MRTK/Tests/TestUtilities/packagetemplate.json:7]:
```suggestion
    "description": "Test Utilities are a set of scripts that accelerate writing tests for features built using the Microsoft Mixed Reality Toolkit",
```

File: UPM/Documentation~/README.md [hunk 2]
```diff
@@ -1,4 +1,3 @@
 # Microsoft Mixed Reality Toolkit
 
-[Getting Started](https://docs.microsoft.com/en-us/windows/mixed-reality/develop/unity/mrtk-getting-started)
-[MRTK Documentation on GitHub](https://microsoft.github.io/MixedRealityToolkit-Unity/README.html)
```
Comment (keveleigh) [inline UPM/Documentation~/README.md:4]: I think there's value in keeping a link to https://docs.microsoft.com/en-us/windows/mixed-reality/develop/unity/mrtk-getting-started. It's a different set of information, and slightly more tailored to "getting started" than the mrtk-unity page

**Comment by Railboy** [comment]
>But I feel it could go even further.

If we want to cut to the bone we could omit definitions for the core services, interfaces and profiles (Camera, Input, Boundary etc). So basically nothing beyond the base service classes found in Core/Services.

With those gone you could eliminate the majority of the interfaces, inspectors, providers etc.

This would necessitate leaving out the MixedRealityToolkit script itself but given that most devs consuming this would want a simple service registry anyway I don't think that would be a problem.

Teasing that all apart is a big project, though. Using a subset of core as-is seems fine in the meantime.

File: Assets/MRTK/StandardAssets/packagetemplate.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
     "msftFeatureCategory": "Mixed Reality Toolkit",
     "version": "%version%",
     "unity": "2019.4",
-    "description": "Mixed Reality Toolkit standard assets, including; shaders, common materials and fonts.",
+    "description": "The Mixed Reality Toolkit Standard Assets is a set of shaders, textures, materials, fonts and other assets used by the Microsoft Mixed Reality Toolkit.",
```
Comment (keveleigh) [inline Assets/MRTK/StandardAssets/packagetemplate.json:7]: nit: inconsistent use of `is` vs `are` across these descriptions. It feels like we should refer to the names as a singular package and maybe drop "The" at the beginning.

**Comment by davidkline-ms** [comment]
> If we want to cut to the bone

One thing we are trying to balance is keeping the number of packages small so we avoid an "explosion" of packages. Discovering the correct balance is the subject of #9612.

If the concern is how much of the code is retained during IL2CPP code stripping, we will be investigating a surgical approach (read: not link.xml) for MRTK v3.

**Comment by davidkline-ms** [comment]
> Is there a plan, probably beyond this PR, to make an inventory of all the pieces of Core and further split them down into self-contained packages?

Please see #9612 for the discussion on granularity.

**Comment by djee-ms** [comment]
> > to work around Unity's arbitrary script execution order
> 
> To be fair...it isn't really arbitrary. If that's all you're looking to do, you're probably better off explicitly setting it at

Having a workaround for an arbitrary default does not make the default any less arbitrary. We _can_ override that arbitrary order for MRTK scripts, but each time we forget to do so some bugs appear. It is also fairly difficult to track dependency chains manually (W needs X to be initialized, which needs Y, which needs Z, ...) and this Script Execution Order doesn't offer a way to highlight/track them. Moreover, we have no control over the interleaving with scripts from users' projects and how they might set themselves up, possibly running even before MRTK ones. The only real robust solution to this is an on-demand lazy initialization like implemented by the current service singleton system of MRTK, whereby any singleton-like object is guaranteed to be initialized just-in-time before it's accessed, no matter what.

File: UPM/Documentation~/README.md (hunk 2 above)
Comment (davidkline-ms) [reply UPM/Documentation~/README.md]: restored

**Comment by davidkline-ms** [comment]
> a popup window will rub folks the wrong way

The latest commit moves the configuration dialog and related code out of Core and into a new location (within the Foundation). While I anticipate this to be the last packaging change for 2.7, more is on the table for v3.0.

**Comment by davidkline-ms** [comment]
Closing this. It will be reevaluated for v3.0

**Pull request closed without merging**